librad/src/profile.rs

```diff
@@ -0,0 +1,211 @@
+/// `..`.
+fn is_valid_profile_id(id: &str) -> bool {
+    let mut components = Path::new(id).components();
+
+    match components.next() {
+        Some(path::Component::Normal(_)) => {},
+        _ => return false,
+    }
+
+    if components.next().is_some() {
+        return false;
+    }
+
+    true
+}
+
+/// Read the profile ID from a file or create the file and write a newly
+/// generated profile ID to it.
+fn load_profile_id(config_dir: &Path) -> Result<String, Error> {
+    let active_profile_path = config_dir.join("active_profile");
+
+    match fs::read_to_string(&active_profile_path) {
+        Ok(content) => {
+            let id = match content.lines().next() {
+                Some(id) => id,
+                None => {
+                    return Err(Error::InvalidProfileIdFromFile {
+                        path: active_profile_path,
+                        id: "".to_string(),
+                    });
+                },
+            };
+
+            if !is_valid_profile_id(id) {
+                return Err(Error::InvalidProfileIdFromFile {
+                    path: active_profile_path,
+                    id: id.to_string(),
+                });
+            }
+
+            Ok(id.to_string())
+        },
+        Err(err) => {
+            if err.kind() == io::ErrorKind::NotFound {
+                let id = Uuid::new_v4().to_hyphenated().to_string();
+                fs::create_dir_all(config_dir)?;
+                fs::write(active_profile_path, &id)?;
+                Ok(id)
+            } else {
+                Err(Error::from(err))
+            }
+        },
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn load_profile_id_test() {
+        let tempdir = tempfile::tempdir().unwrap();
+
+        let id1 = load_profile_id(tempdir.path()).unwrap();
+        let id2 = load_profile_id(tempdir.path()).unwrap();
+        assert_eq!(id2, id1);
+        fs::remove_dir_all(tempdir.path()).unwrap();
+
+        let id3 = load_profile_id(tempdir.path()).unwrap();
+        assert_ne!(id3, id1);
+    }
+
+    #[test]
+    fn invalid_profile_id_from_file() {
+        let tempdir = tempfile::tempdir().unwrap();
+
+        let profile_id_path = tempdir.path().join("active_profile");
+        let invalid_id = "foo/bar";
+        std::fs::write(profile_id_path, invalid_id).unwrap();
+
+        let result = load_profile_id(tempdir.path());
+        assert!(matches!(
+            result,
+            Err(Error::InvalidProfileIdFromFile { .. })
+        ));
+    }
+
+    #[test]
+    fn load_profile_strip() {
+        let tempdir = tempfile::tempdir().unwrap();
+
+        let profile_id_path = tempdir.path().join("active_profile");
+        let content = "foo\n";
+        std::fs::write(profile_id_path, content).unwrap();
+        let id = load_profile_id(tempdir.path()).unwrap();
+        assert_eq!(id, "foo");
+    }
+
+    #[test]
+    fn empty_profile() {
```

Comment:
This doesn't seem to be testing the empty case, it looks like the test case above? :eyes:

Reply:
Fxied